pkg/vault/instances.go

```diff
@@ -245,60 +271,68 @@ func createClient(addr string,
-			fmt.Println(fmt.Sprintf("SKIPPING ALL RECONCILIATION FOR: %s\n", addr))
-			return // skip entire reconcilation for this instance
+			log.WithError(err).Errorf("[Vault Client] failed to login to `%s` with kube sa token", addr)
+			log.Warnf("SKIPPING ALL RECONCILIATION FOR: %s", addr)
+			return // Skip entire reconciliation for this instance.
 		}
 	} else {
 		accessCreds := make(map[string]string)
 		for _, cred := range bundle.VaultSecrets {
 			// masterAddress hard-coded because all "child" vault access credentials must be pulled from master
 			processedCred, err := GetVaultSecretField(masterAddress, cred.Path, cred.Field, bundle.SecretEngine)
 			if err != nil {
-				log.WithError(err).Fatal()
+				log.WithError(err).Fatal("[Vault Client] unable to retrieve credentials from master Vault")
 			}
 			accessCreds[cred.Name] = processedCred
 		}
 
 		// at minimum, one element will exist in secrets regardless of type
 		// type is same across all VaultSecrets associated with a particular instance address
-		var token string
+		var clientToken string
 		switch bundle.VaultSecrets[0].Type {
 		case APPROLE_AUTH:
-			t, err := client.Logical().Write("auth/approle/login", map[string]interface{}{
-				"role_id":   accessCreds[ROLE_ID],
-				"secret_id": accessCreds[SECRET_ID],
+			err := utils.Retry(defaultTokenRetryAttempts, defaultTokenRetrySleep, func() error {
+				secret, err := client.Logical().Write("auth/approle/login", map[string]interface{}{
+					"role_id":   accessCreds[ROLE_ID],
+					"secret_id": accessCreds[SECRET_ID],
+				})
+				if err != nil {
+					return utils.RetryStop(err)
+				}
+				// Writing to the AppRole mount endpoint to "login" to obtain a new
+				// token might return an empty response without necessarily failing.
+				if secret.Auth == nil || secret.Auth.ClientToken == "" {
+					log.Warn("[Vault Client] received empty authentication information. Retrying...")
+					return errClientTokenInvalid
+				}
+				clientToken = secret.Auth.ClientToken
+				return nil
 			})
 			if err != nil {
-				log.WithError(err)
-				fmt.Println(fmt.Sprintf("[Vault Client] failed to login to %s with AppRole credentials", addr))
-				fmt.Println(fmt.Sprintf("SKIPPING ALL RECONCILIATION FOR: %s\n", addr))
-				return // skip entire reconcilation for this instance
+				log.WithError(err).Errorf("[Vault Client] failed to login to `%s` with AppRole credentials", addr)
+				log.Warnf("SKIPPING ALL RECONCILIATION FOR: %s", addr)
+				return // Skip entire reconciliation for this instance.
 			}
-			token = t.Auth.ClientToken
 		case TOKEN_AUTH:
-			token = accessCreds[TOKEN]
+			clientToken = accessCreds[TOKEN]
 		}
-		client.SetToken(token)
+		client.SetToken(clientToken)
 	}
 
 	// test client
 	_, err = client.Sys().ListAuth()
 	if err != nil {
-		log.WithError(err)
-		fmt.Println(fmt.Sprintf("[Vault Client] failed to login to %s", addr))
-		fmt.Println(fmt.Sprintf("SKIPPING ALL RECONCILIATION FOR: %s\n", addr))
```

Comment:
@dwelch0, I was curious why you did it like this. Was there some specific requirement for how the message is printed or formatted? I may be breaking something here with my change. Let me know. :smile:

Reply:
No real motivation for it :sweat_smile: . There should be no issues with changing it to whatever format you prefer.

Reply:
@dwelch0, no worries! I wanted to make sure that I am not going to break something.